File: backend/src/main/java/com/borisov/alexander/pokepedia/service/PokepediaService.java
package com.borisov.alexander.pokepedia.service;
import com.borisov.alexander.pokepedia.model.Pokemon;
import com.borisov.alexander.pokepedia.repository.PokemonRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PokepediaService {

    private final PokemonRepository pokemonRepository;

    public PokepediaService(PokemonRepository pokemonRepository) {
        this.pokemonRepository = pokemonRepository;
    }

    /**
     * Returns a list of all 899 pokemons.
     * @param NONE
     * @return List<Pokemon>
     */
    public List<Pokemon> findAllPokemons() {
        return pokemonRepository.findAll();
    }

    /**
     * Returns a list of pokemon(s) filtered by their type(s) in the select boxes.
     * @param List<String>
     * @return List<Pokemon>
     */
    public List<Pokemon> findPokemonsByTypeIn(List<String> type) {
        return pokemonRepository.findPokemonsByTypeIn(type);
    }

    /**
     * Returns a list of pokemon(s) filtered by their name(s) in the input field.
     * @param name
     * @return List<Pokemon>
     */
    public List<Pokemon> findPokemonsByName(String name) {
        return pokemonRepository.findPokemonsByName(name);
    }
}

File: frontend/pokepediafrontend/src/app/service/pokepedia.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Pokemon } from '../model/pokemon';
import { environment } from 'src/environments/environment';

@Injectable({ providedIn: 'root' })
export class PokepediaService {
  constructor(private http: HttpClient) {}

  // URL for HTTP requests.
  private apiServerUrl = environment.apiBaseUrl;

  // Array of the selected pokemon types.
  private selectedTypes: string[] = [];

  /**
   * HTTP GET request of all pokemons from the database.
   * @returns Pokemon[] array of pokemons.
   */
  public getPokemons(): Observable<Pokemon[]> {
    return this.http.get<Pokemon[]>(`${this.apiServerUrl}/pokepedia/all`);
  }

  /**
   * HTTP GET request of a certain pokemon(s) type which is(are) selected in the <select> checkbox.
   * @param string[] type.
   * @returns Pokemon[] array of pokemons.
   */
  public getPokemonsByType(type: string[]): Observable<Pokemon[]> {
    return this.http.get<Pokemon[]>(
      `${this.apiServerUrl}/pokepedia/type?t=${type}`
    );
  }

  /**
   * HTTP GET request of a pokemon(s) by name which is entered in the input field.
   * @param string name.
   * @returns Pokemon[] array of pokemons.
   */
  public getPokemonsByName(name: string): Observable<Pokemon[]> {
    return this.http.get<Pokemon[]>(
      `${this.apiServerUrl}/pokepedia/name?n=${name}`
    );
  }

  /**
   * Append selected checkbox(ex) type(s) to the array in order to make further HTTP GET request.
   * @param string currentType.
   * @returns string[] array of selected types.
   */
  public appendType(currentType: string): string[] {
    let index: number;
    if (this.selectedTypes.includes(currentType)) {
      index = this.selectedTypes.indexOf(currentType);
      this.selectedTypes.splice(index, 1);
      return this.selectedTypes;
    } else {
      this.selectedTypes.push(currentType);
      return this.selectedTypes;
    }
  }
}
